# Workflow Integration in Jupyter

Research projects are often involve complex processes with multiple steps that are interconnected. 
Each of these steps might focus on a specific part of the analysis, such as data cleaning, model building, or visualization.
While these steps can be developed and run separately, they are usually not independent; one step may depend on the outcome of a previous step.
For instance, before you can build a model, you might need results from data preprocessing or specific configuration settings.

In Jupyter notebooks, there are various scenarios where one notebook might need to use content from another notebook.
Sometimes, you might want to run an entire notebook to make sure everything works in sequence.
At other times, you may only need to pull in specific parts of another notebook, such as a function, a configuration setting, or a block of code.
Jupyter provides several built-in and external tools to handle these scenarios efficiently.
In this notebook, we will explore how to use these tools to manage complex workflows such that notebooks can either run entire sequences or access just the specific content they need, making the research process more efficient and organized.

We start by looking at magic commands to manage notebooks, how to sequence of notebooks from another notebook, how to import contents of notebook like it was a script, and end by practicing setting up a configuration notebook. 


## Magic commands

Magic commands in Jupyter notebooks are special commands that provide a convenient way to perform certain tasks and control the behavior of the Jupyter environment. They start with a % (for line magics, which affect a single line) or %% (for cell magics, which affect the entire cell). These commands are designed to facilitate tasks such as running scripts, profiling code, timing execution, and integrating with other environments or languages.

When we are developing an code as researchers, we are mainly concerned with how long an analysis runs and how to store and load results. In this section, let's focus on some of Jupyter magic commands that let's us do just that.

**Example** Measure how long it takes to sum numbers from 1 to 100

In [2]:
%timeit sum(range(100))

651 ns ± 11.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


You might be surprised that it took longer than to just run the code. That is because %timeit runs the line of code multiple times with a lot of iterations in each run to get a reliable estimation of the execution time. The output you see here will look something like this

`651 ns ± 11.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)`

Here's what it means:

| **Number/Value**      | **Explanation**                                                                                                                                   |
|-----------------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| `651 ns`              | The mean (average) execution time of the code per loop. The unit here is nanoseconds (ns), meaning it took an average of 651 nanoseconds per loop. |
| `± 11.4 ns`           | The standard deviation of the execution time. It indicates the variability in the execution time across different runs. In this case, it varies by 11.4 nanoseconds. |
| `7 runs`              | The number of times `%timeit` ran the test. In this case, the test was executed 7 times to gather the statistics for the mean and standard deviation. |
| `1,000,000 loops each`| The number of iterations (loops) that were executed per run. Here, the code was repeated 1,000,000 times in each of the 7 runs to get a more accurate timing. |


Measure how long it takes to sum numbers from 1 to 1000

In [3]:
%timeit sum(range(1000))

15 μs ± 426 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Let's take this further. 
Measure how long it takes to sum numbers from 1 to 100000

In [4]:
%timeit sum(range(100000))

1.71 ms ± 32.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Do you notice that number of loops reduced as the code took longer to execute? This is because %timeit automatically adjusts the number of loops to ensure an accurate measurement based on the how quickly the code being tested runs. 

**Example** Measure how long it takes to sum numbers from 1 to 100 with 1000 loops per iteration

In [5]:
%timeit -n 1000 sum(range(100))

745 ns ± 11.2 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Measure how long it takes to sum numbers from 1 to 100 with 10000 loops per iteration

In [6]:
%timeit -n 10000 sum(range(100))

746 ns ± 158 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


Measure how long it takes to sum numbers from 1 to 100 with 5 iterations.

Hint: Use -r to set number of iterations

In [7]:
%timeit -r 5 sum(range(100))

564 ns ± 9.04 ns per loop (mean ± std. dev. of 5 runs, 1,000,000 loops each)


Measure how long it takes to sum numbers from 1 to 100 with 5 iterations and 1000 loops per iteration

In [8]:
%timeit -r 5 -n 1000 sum(range(100))

533 ns ± 153 ns per loop (mean ± std. dev. of 5 runs, 1,000 loops each)


In the below exercises, let's practice loading contents of another file. 

**Example** Load contents of `text_config` file

In [11]:
%load data/text_config.txt

As soon as you execute the cell with %load, the following happens

1. %load command itself is turned into a comment
2. Below the comment, the contents of the file gets loaded
3. You can edit it and then execute the line

Load contents of `python_config` file

In [13]:
# %load data/python_config.py
num_exp=10 # number of experiments
scientist="John Doe" # name of the scientist

It can also work with other notebooks. It loads the whole notebook.

Load contents of notebook_config.file

In [14]:
# %load data/notebook_config.ipynb
{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4241faf6-feb5-404f-a227-369aa7057d93",
   "metadata": {},
   "source": [
    "These values were changed on 2024-10-11\n",
    "\n",
    "Parameter `num_exp` describes how many experiments were conducted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c20776fc-d2bf-4e4c-933d-6a0816dc0492",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_exp = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41be7a78-31b2-47de-8fc4-22fd6867e29c",
   "metadata": {},
   "source": [
    "Parameter `scientist` is the name of the person who conducted the series of experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cb3b0e40-77ad-4911-8671-ee719c2da1cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "scientist = \"John Doe\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}


**Example** Store "Mice Visual Cortex Analysis" as `project_name` so any notebook can access it

In [17]:
project_name = 'Mice Visual Cortex Analysis'
%store project_name

Stored 'project_name' (str)


Now this variable is stored on the disk in ~/.ipython. It will be available to all jupyter notebooks as long as the notebooks are run in same environment

Store 123456 as `project_id` so any notebook can access it

In [19]:
project_id = 123456
%store project_id

Stored 'project_id' (int)


Store "Genius Lab" as `lab_name` so any notebook can access it.

In [20]:
lab_name = "Genius lab"
%store lab_name

Stored 'lab_name' (str)


**Example** In another notebook, retrieve `project_name`

In [22]:
%store -r project_name
project_name

'Mice Visual Cortex Analysis'

In another notebook, retrieve project_id

In [25]:
%store -r project_id
project_id

123456

In another notebook, retrieve lab_name

In [26]:
%store -r lab_name
lab_name

'Genius lab'

So far we only saw magic commands that run on a single line of code. Let's look into some cell magic commands that run on entire cell

**Example** Measure how long it takes to sum numbers upto 1000 with loop. 

In [30]:
%%time
result = 0
for i in range(1000):
    result += i

CPU times: total: 0 ns
Wall time: 0 ns


Measure how long it takes to sum numbers upto 10000000 with loop.

In [31]:
%%time
result = 0
for i in range(10000000):
    result += i

CPU times: total: 500 ms
Wall time: 1.07 s


You can also use it for single line of code as long as the code is below %%time

Measure how long it takes to sum numbers upto 10000000 without loop.

In [33]:
%%time 
sum(range(10000000))

CPU times: total: 46.9 ms
Wall time: 183 ms


49999995000000

We might also want to write contents of a single cell into a file of its own. This can be useful when you write functions or have a list of variables that you want to store as a python script to access later on.

**Example** Store `experiment_name`, `num_mice`, `num_neuropixels` in a file called `experiment_info.txt`

In [34]:
%%writefile experiment_info.txt
experiment_name = "Mice Visual Cortex"
num_mice = 25
num_neuropixels = 300

Writing experiment_info.txt


Store `experiment_name`, `num_mice`, `num_neuropixels` in a file called `experiment_info.py`

In [35]:
%%writefile experiment_info.py
experiment_name = "Mice Visual Cortex"
num_mice = 25
num_neuropixels = 300

Writing experiment_info.py


Add `num_electrodes` to `experiment_info.py`

Hint: Use -a

In [37]:
%%writefile -a experiment_info.py
num_electrodes = 100

Appending to experiment_info.py


## Run a whole notebook from here %run

## Access only configurations from another notebook

## import_ipynb module

.